trees/treeconfig.go

```diff
@@ -0,0 +1,185 @@
+func (t *Tree) InitializeWithPIDs(ids []hotstuff.ID) error {
+	if len(ids) != t.configurationSize {
+		return fmt.Errorf("invalid number of replicas")
+	}
+	// check for duplicate IDs
+	idToIndexMap := make(map[hotstuff.ID]int)
+	for index, id := range ids {
+		if _, ok := idToIndexMap[id]; !ok {
+			idToIndexMap[id] = index
+		} else {
+			return fmt.Errorf("duplicate replica ID: %d", id)
+		}
+	}
+	t.posToIDMapping = ids
+	return nil
+}
+
+func (t *Tree) GetTreeHeight() int {
+	return t.height
+}
+
+func (t *Tree) getPosition() (int, error) {
+	pos := slices.Index(t.posToIDMapping, t.id)
+	if pos == -1 {
+		return 0, fmt.Errorf("replica not found")
+	}
+	return pos, nil
+}
+
+func (t *Tree) getReplicaPosition(replicaId hotstuff.ID) (int, error) {
+	pos := slices.Index(t.posToIDMapping, replicaId)
+	if pos == -1 {
+		return 0, fmt.Errorf("replica not found")
+	}
+	return pos, nil
+}
+
+// GetParent fetches the ID of the parent, if root, returns itself.
+func (t *Tree) GetParent() (hotstuff.ID, bool) {
+	myPos, err := t.getPosition()
+	if err != nil {
+		return hotstuff.ID(0), false
+	}
+	if myPos == 0 {
+		return t.id, false
+	}
+	return t.posToIDMapping[(myPos-1)/t.branchFactor], true
+}
+
+// GetChildren returns the children of the replicas, if any.
+func (t *Tree) GetChildren() []hotstuff.ID {
+	return t.GetChildrenOfNode(t.id)
+}
+
+func (t *Tree) isWithInIndex(position int) bool {
+	return position < t.configurationSize
+}
+
+// IsRoot return true if the replica is at root of the tree.
+func (t *Tree) IsRoot(nodeID hotstuff.ID) bool {
+	pos, err := t.getReplicaPosition(nodeID)
+	if err != nil {
+		return false
+	}
+	return pos == 0
+}
+
+// GetChildrenOfNode returns the children of a specific replica.
+func (t *Tree) GetChildrenOfNode(nodeID hotstuff.ID) []hotstuff.ID {
+	children := make([]hotstuff.ID, 0)
+	nodePos, err := t.getReplicaPosition(nodeID)
+	if err != nil {
+		return children
+	}
+	for i := 1; i <= t.branchFactor; i++ {
+		childPos := (t.branchFactor * nodePos) + i
+		if t.isWithInIndex(childPos) {
+			children = append(children, t.posToIDMapping[childPos])
+		} else {
+			break
+		}
+	}
+	return children
+}
+
+// getHeight returns the height of a given replica.
+func (t *Tree) getHeight(nodeID hotstuff.ID) int {
+	if t.IsRoot(nodeID) {
+		return t.height
+	}
+	nodePos, err := t.getReplicaPosition(nodeID)
+	if err != nil {
+		return 0
+	}
+	startLimit := 0
+	endLimit := 0
+	for i := 1; i < t.height; i++ {
+		startLimit = startLimit + int(math.Pow(float64(t.branchFactor), float64(i-1)))
+		endLimit = endLimit + int(math.Pow(float64(t.branchFactor), float64(i)))
+		if nodePos >= startLimit && nodePos <= endLimit {
+			return t.height - i
+		}
+	}
+	return 0
+}
+
+// GetHeight returns the height of the replica
+func (t *Tree) GetHeight() int {
+	return t.getHeight(t.id)
+}
+
+// GetPeers returns the peers of given ID, if any.
+func (t *Tree) GetPeers(nodeID hotstuff.ID) []hotstuff.ID {
```

Comment:
Let's rename it to: `PeersOf(id hotstuff.ID)`. Or perhaps `NeighborsOf`?

Is this meant to return both the `parent` and the `children`? Right now, I think it only returns the children.

And if `nodeID` is the root, shouldn't this return its children?

Reply:
It only returns the siblings of the replica, along with itself